File: server/routes/misc.js
import express from 'express';
import fs from 'fs';
import path from 'path';
import { db } from '../db.js';

const router = express.Router();

router.get('/status', (req, res) => res.json({ message: 'EduTrack API Connected', version: '2.0.0' }));

router.get('/notifications/user/:id', (req, res) => {
    const userId = parseInt(req.params.id);
    const userNotifications = (db.notifications || []).filter(n => n.user_id === userId);
    res.json(userNotifications.sort((a, b) => new Date(b.created_at) - new Date(a.created_at)));
});

router.post('/notifications/:id/read', (req, res) => {
    const id = parseInt(req.params.id);
    const updated = db.update('notifications', id, { status: 'read' });
    if (updated) {
        res.json({ message: 'Marked as read' });
    } else {
        res.status(404).json({ error: 'Not found' });
    }
});

router.post('/notifications/mark-all-read/:userId', (req, res) => {
    const userId = parseInt(req.params.userId);
    const data = JSON.parse(fs.readFileSync(path.join(process.cwd(), 'server', 'database.json'), 'utf-8'));
    data.notifications = (data.notifications || []).map(n => 
        n.user_id === userId ? { ...n, status: 'read' } : n
    );
    fs.writeFileSync(path.join(process.cwd(), 'server', 'database.json'), JSON.stringify(data, null, 4));
    res.json({ message: 'Marked all as read' });
});

router.post('/notifications/clear-all/:userId', (req, res) => {
    const userId = parseInt(req.params.userId);
    // Since we don't have a bulk delete in db.js, we do it manually and save
    const data = JSON.parse(fs.readFileSync(path.join(process.cwd(), 'server', 'database.json'), 'utf-8'));
    data.notifications = (data.notifications || []).filter(n => n.user_id !== userId);
    fs.writeFileSync(path.join(process.cwd(), 'server', 'database.json'), JSON.stringify(data, null, 4));
    res.json({ message: 'Cleared all' });
});

router.get('/announcements', (req, res) => {
    const announcements = (db.announcements || []).sort((a, b) => new Date(b.date) - new Date(a.date));
    res.json(announcements);
});

router.post('/announcements', (req, res) => {
    const { title, content, priority, schoolId, userId, target } = req.body;
    const newAnnouncement = {
        id: db.getNextId('announcements'),
        school_id: schoolId || 1,
        title,
        content,
        priority: priority || 'normal',
        target: target || 'all',
        date: new Date().toISOString().split('T')[0]
    };
    db.push('announcements', newAnnouncement);
    
    // Create notifications for selected target
    let userFilter = [];
    if (target === 'teachers') userFilter = ['Teacher'];
    else if (target === 'students') userFilter = ['Student'];
    else userFilter = ['Teacher', 'Student'];

    const users = db.users.filter(u => userFilter.includes(u.role));
    users.forEach(u => {
        db.push('notifications', {
            id: db.getNextId('notifications'),
            user_id: u.id,
            title: `Announcement: ${title}`,
            message: content,
            type: 'announcement',
            status: 'unread',
            created_at: new Date().toISOString()
        });
    });

    db.auditLog('BROADCAST_ANNOUNCEMENT', userId || 1, { title, target });
    res.json(newAnnouncement);
});

export default router;

File: backend/api/views/misc.py
from rest_framework.views import APIView
from rest_framework.response import Response
from rest_framework.permissions import AllowAny, IsAuthenticated

from api.models import User, Announcement, Notification
from api.utils import create_audit_log, create_notification


class StatusView(APIView):
    permission_classes = [AllowAny]

    def get(self, request):
        return Response({'message': 'EduTrack API Connected', 'version': '3.0.0-django'})


class UserNotificationsView(APIView):
    permission_classes = [IsAuthenticated]

    def get(self, request, user_id):
        notifications = Notification.objects.filter(user_id=int(user_id)).order_by('-created_at')
        return Response([{
            'id': n.id, 'user_id': n.user_id, 'title': n.title,
            'message': n.message, 'type': n.type, 'status': n.status,
            'created_at': n.created_at.isoformat(),
        } for n in notifications])


class MarkNotificationReadView(APIView):
    permission_classes = [IsAuthenticated]

    def post(self, request, notification_id):
        try:
            n = Notification.objects.get(id=int(notification_id))
            n.status = 'read'
            n.save(update_fields=['status'])
            return Response({'message': 'Marked as read'})
        except Notification.DoesNotExist:
            return Response({'error': 'Not found'}, status=404)


class MarkAllNotificationsReadView(APIView):
    permission_classes = [IsAuthenticated]

    def post(self, request, user_id):
        Notification.objects.filter(user_id=int(user_id)).update(status='read')
        return Response({'message': 'Marked all as read'})


class ClearAllNotificationsView(APIView):
    permission_classes = [IsAuthenticated]

    def post(self, request, user_id):
        Notification.objects.filter(user_id=int(user_id)).delete()
        return Response({'message': 'Cleared all'})


class AnnouncementListView(APIView):
    permission_classes = [AllowAny]

    def get(self, request):
        announcements = Announcement.objects.all()
        return Response([{
            'id': a.id, 'title': a.title, 'content': a.content,
            'priority': a.priority, 'target': a.target,
            'section_id': a.section_id,
            'date': str(a.date), 'school_id': 1,
        } for a in announcements])


class AnnouncementCreateView(APIView):
    permission_classes = [IsAuthenticated]

    def post(self, request):
        title = request.data.get('title', '')
        content = request.data.get('content', '')
        priority = request.data.get('priority', 'normal')
        target = request.data.get('target', 'all')

        announcement = Announcement.objects.create(
            title=title, content=content, priority=priority, target=target,
            author=request.user
        )

        # Notify target users
        if target == 'teachers':
            users = User.objects.filter(role='Teacher')
        elif target == 'students':
            users = User.objects.filter(role='Student')
        else:
            users = User.objects.filter(role__in=['Teacher', 'Student'])

        for u in users:
            create_notification(u, f'Announcement: {title}', content, 'announcement')

        create_audit_log('BROADCAST_ANNOUNCEMENT', request.user, title=title, target=target)

        return Response({
            'id': announcement.id, 'title': announcement.title,
            'content': announcement.content, 'priority': announcement.priority,
            'target': announcement.target, 'date': str(announcement.date), 'school_id': 1,
        })
